### PR TITLE
chore: treat symlink as file when adding to zip

### PR DESCRIPTION
**Pull request by j3tm0t0**
This commit modifies the `addToZip` function in `archive.go` to treat symlinks as files. If a file is a symlink, the function reads the symlink target and retrieves the file information for the target. This ensures that the correct file information is used when adding the file to the zip archive.

Refactor the code to handle symlinks and update the `path` variable accordingly before creating the zip file header.

This change improves the behavior of the `addToZip` function when dealing with symlinks, ensuring that the correct file information is used for symlinks in the zip archive.

This is useful when you have common files shared by multiple functions like below  but do not want to maintain it as layer.

```
.
├── common
│   └── hoge.py
├── lambda
│   ├── function.json
│   ├── hoge.py -> ../common/hoge.py
│   └── lambda_function.py
└── lambda2
    ├── function.json
    ├── hoge.py -> ../common/hoge.py
    └── lambda_function.py
```

Without this patch, zip archive is corrupt and cannot be properly decompressed. 

```
~/lambroll-symlink/lambda$ lambroll archive
2024/06/12 19:13:50 [info] lambroll v1.0.4
2024/06/12 19:13:50 [info] creating zip archive from .
2024/06/12 19:13:50 [info] zip archive wrote 1196 bytes
2024/06/12 19:13:50 [info] writing zip archive to function.zip
~/lambroll-symlink/lambda$ cd ..
~/lambroll-symlink$ unzip lambda/function.zip
Archive:  lambda/function.zip
  inflating: .python-version
  inflating: .vscode/launch.json
  inflating: hoge.py                 -> def hoge():^J    return "hoge"^J
  inflating: lambda_function.py
finishing deferred symbolic links:
  hoge.py                -> def hoge():^J    return "hoge"^J
```

but can be decompressed with this patch.

```
~/lambroll-symlink/lambda$ ~/bin/lambroll archive
2024/06/12 19:18:48 [info] lambroll
2024/06/12 19:18:48 [info] creating zip archive from .
2024/06/12 19:18:48 [info] zip archive wrote 1196 bytes
2024/06/12 19:18:48 [info] writing zip archive to function.zip
~/lambroll-symlink/lambda$ cd ..
~/lambroll-symlink$ unzip lambda/function.zip
Archive:  lambda/function.zip
  inflating: .python-version
  inflating: .vscode/launch.json
  inflating: hoge.py
  inflating: lambda_function.py
```

### CODE REVIEW

File: archive.go [hunk 1]
```diff
@@ -118,6 +118,20 @@ func matchExcludes(path string, excludes []string) bool {
...
 
 func addToZip(z *zip.Writer, path, relpath string, info os.FileInfo) error {
+	// treat symlink as file
+	if info.Mode()&os.ModeSymlink != 0 {
+		linkTarget, err := os.Readlink(path)
+		if err != nil {
+			log.Printf("[error] failed to read symlink %s: %s", path, err)
+			return err
+		}
+		info, err = os.Stat(linkTarget)
```
Comment (fujiwara) [inline archive.go:128]: `lambroll archive --src src/` raises an error "failed to stat symlink target" here.
Resolving the relative path from the `path` is needed if the symlink is relative.

**Review by fujiwara — COMMENTED** [review]
@j3tm0t0 Thank you!
I commented. Confirm it, please.

And I think lambroll should keep symlinks links to the file in the archive. Because it might be the cause of increasing zip file size unnecessarily.

**Comment by j3tm0t0** [comment]
> @j3tm0t0 Thank you! I commented. Confirm it, please.

thank you for your comment! I never used `--src` option before and could not recognize the problem. I will try to fix it.

> And I think lambroll should keep symlinks links to the file in the archive. Because it might be the cause of increasing zip file size unnecessarily.

The current implementation is just broken: it treats a symlink as a symlink but the symlink is not pointing the target properly but the content of the target file is unintentionally considered as symlink target info (and it is not working as expected).

I think solution could be either of these:
1. treat symlinks as files, and content will be copied in the zip file (zip command does so by default)
2. treat symlinks as symlinks (like zip command does with -y/--symlink option) // in this case, only files under source directory can be relatively referred

Honestly speaking, I would like the option 1 as the default behavior (as zip command does so by default), but at least I want the lambroll has an option to do so with some extra option arg.

Could you consider it once I could resolve the `--src` option problem, please?

**Comment by fujiwara** [comment]
@j3tm0t0 I added a commit to this PR by #422. Thank you!

**Pull request merged**